adapters/pangle/pangle.go

```diff
@@ -0,0 +1,196 @@
+	*adapters.ExtImpBidder
+	AdType int `json:"adtype,omitempty"`
+}
+
+type pangleBidExt struct {
+	Pangle *bidExt `json:"pangle,omitempty"`
+}
+
+type bidExt struct {
+	AdType *int `json:"adtype,omitempty"`
+}
+
+/* Builder */
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		Endpoint: config.Endpoint,
+	}
+
+	return bidder, nil
+}
+
+/* MakeRequests */
+
+func getAdType(imp openrtb.Imp, parsedImpExt *wrappedExtImpBidder) int {
+	// video
+	if imp.Video != nil {
+		if parsedImpExt != nil && parsedImpExt.Prebid != nil && parsedImpExt.Prebid.IsRewardedInventory == 1 {
+			return 7
+		}
+		if imp.Instl == 1 {
+			return 8
+		}
+	}
+	// banner
+	if imp.Banner != nil {
+		if imp.Instl == 1 {
+			return 2
+		} else {
+			return 1
+		}
+	}
+	// native
+	if imp.Native != nil && len(imp.Native.Request) > 0 {
+		return 5
+	}
+
+	return -1
+}
+
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var requests []*adapters.RequestData
+	var errs []error
+
+	requestCopy := *request
+	for _, imp := range request.Imp {
+		var impExt wrappedExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			errs = append(errs, fmt.Errorf("failed unmarshalling imp ext (err)%s", err.Error()))
+			continue
+		}
+		// detect and fill adtype
+		if adType := getAdType(imp, &impExt); adType == -1 {
+			errs = append(errs, adapters.BadInput("not a supported adtype"))
+			continue
+		} else {
+			impExt.AdType = adType
+			if newImpExt, err := json.Marshal(impExt); err == nil {
+				imp.Ext = newImpExt
+			} else {
+				errs = append(errs, fmt.Errorf("failed re-marshalling imp ext with adtype"))
+				continue
+			}
+		}
+		// for setting token
+		var bidderImpExt openrtb_ext.ImpExtPangle
+		if err := json.Unmarshal(impExt.Bidder, &bidderImpExt); err != nil {
+			errs = append(errs, fmt.Errorf("failed unmarshalling bidder imp ext (err)%s", err.Error()))
+			continue
+		}
+
+		requestCopy.Imp = []openrtb.Imp{imp}
+		requestJSON, err := json.Marshal(requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		requestData := &adapters.RequestData{
+			Method: "POST",
+			Uri:    a.Endpoint,
+			Body:   requestJSON,
+			Headers: http.Header{
+				"TOKEN":        []string{bidderImpExt.Token},
+				"Content-Type": []string{"application/json"},
+			},
+		}
+		requests = append(requests, requestData)
+	}
+
+	return requests, errs
+}
+
+/* MakeBids */
+
+func getMediaTypeForBid(bid *openrtb.Bid) (openrtb_ext.BidType, error) {
+	if bid == nil {
+		return "", fmt.Errorf("the bid request object is nil")
+	}
+
+	var bidExt pangleBidExt
+	if err := json.Unmarshal(bid.Ext, &bidExt); err != nil {
+		return "", fmt.Errorf("invalid bid ext")
+	} else if bidExt.Pangle == nil || bidExt.Pangle.AdType == nil {
+		return "", fmt.Errorf("missing pangleExt/adtype in bid ext")
```

Comment:
Could you please add a test for this condition?

Reply:
added